TypeCobol/Compiler/CodeModel/SymbolTable.cs

```diff
@@ -224,59 +244,95 @@ public List<DataDefinition> GetVariablesByType(DataType dataType, IEnumerable<Da
+            if (variable.DataType == dataType)
             {
-                if(!foundedVariables.Any(v => v == variable))
-                    foundedVariables.Add(variable);
-                return;
+                return true;
             }
 
-            if (variable.DataType != null && variable.DataType != DataType.Boolean && variable.DataType.CobolLanguageLevel > CobolLanguageLevel.Cobol85)
+            if (variable.TypeDefinition != null)
             {
-                var types = GetTypes(t => t.DataType == variable.DataType, Scope.Intrinsic);
-
-                foreach (var type in types)
+                if (dataType.CobolLanguageLevel == CobolLanguageLevel.Cobol85)
                 {
-                    if (type.Children != null && type.Children.Count > 0)
+                    if (variable.PrimitiveDataType == dataType)
                     {
-                        foreach (var childrenType in type.Children)
-                        {
-                            if (childrenType is DataDefinition && childrenType.Name != null)
-                            {
-                                SeekVariableType(dataType, childrenType as DataDefinition, ref foundedVariables);
-                            }
-                        }
+                        return true;
+                    }
+                }
+                else
+                {
+                    //For type DataType.Name depends on how is written the reference to the type : Prefixed with the Program or not
+                    //TODO use resolved TypeDefinition to compare Type
+                    if (Regex.Match(variable.DataType.Name, @"\b" + dataType.Name + @"\b", RegexOptions.IgnoreCase).Success)
+                    {
+                        return true;
                     }
                 }
             }
 
+            return false;
+        }
 
-            if (variable.Children != null && variable.Children.Count > 0)
+        /// <summary>
+        /// Search all children variables recursively in the specified "variable" that match the specified "dataType".
+        /// </summary>
+        /// <param name="dataType"></param>
+        /// <param name="variable"></param>
+        /// <param name="foundVariables"></param>
+        private void SeekVariableType(DataType dataType, DataDefinition variable, ISet<DataDefinition> foundVariables)
+        {
+            if (DataTypeMatch(dataType, variable))
             {
-                foreach (var children in variable.Children)
+                foundVariables.Add(variable);
+                return; //If a variable is matched don't search its children
+            }
+
+            if (variable.TypeDefinition != null && variable.TypeDefinition != DataType.BooleanType)
+            {
+                SearchChildren(variable.TypeDefinition.Children);
+            }
+
+            SearchChildren(variable.Children);
```

Comment:
I think it is:

```csharp
if (variable.TypeDefinition != null)
{
    if (variable.TypeDefinition != DataType.BooleanType)
    {
        SearchChildren(variable.TypeDefinition.Children);
    }
}
else
{
    SearchChildren(variable.Children);
}
```

Because when a variable has a TypeDef it shouldn't have its own children right ?

Reply:
Yes a typed variable cannot have children. Code fixed.